invisible_cities/cities/detsim_waveforms.py

```
@@ -0,0 +1,74 @@
+import numpy as np
+from warnings import warn
+
+##################################
+######### WAVEFORMS ##############
+##################################
+def create_waveform(times    : np.ndarray,
+                    pes      : np.ndarray,
+                    bins     : np.ndarray,
+                    nsamples : int) -> np.ndarray:
+    """
+    This function builds a waveform from a set of (buffer_time, pes) values.
+    This set is of values come from the times and pes arguments.
+
+    Input
+    times: a vector with the buffer times in which a photoelectron is produced in the
+    detector.
+    pes (from photoelecton): a vector with the photoelectrons produced in the detector in
+    each of the buffer times in times argument.
+    bins: a vector with the output waveform bin times (for example [0, 25, 50, ...] if
+    the detector has a sampling time of 25).
+    nsamples: an integer that controlls the distribution of the photoelectrons in each of
+    the waveform bins. The counts (N) in a given time bin (T) are distributed in the waveform
+    such as the nsamples posterior to T would have N/nsamples counts (included T).
+    nsamples must be >=1 an <len(bins).
+    """
+    if nsamples<1:
+        raise Exception("nsamples<1. nsamples must be >= 1")
+    if nsamples>len(bins):
+        warn("nsamples > number of bins. Shrinking it to number of bins")
+        nsamples = len(bins)
```

Exceeding the number of bins doesn't sound like a mild thing. I would also raise an exception here.

---

You can combine both checks into a single one